rust/ast/src/associated_data.rs

```diff
@@ -0,0 +1,126 @@
+use crate::SourceLocation;
+use std::collections::HashMap;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
+struct Key {
+    type_id: NodeTypeId,
+    loc: SourceLocation,
+}
+
+impl Key {
+    fn new<NodeT>(node: &NodeT) -> Self
+    where
+        NodeT: SourceLocationAccessor + NodeTypeIdAccessor,
+    {
+        Self {
+            type_id: node.get_type_id(),
+            loc: node.get_loc(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Item<ValueT> {
+    key: Key,
+
+    // Using `Option` to take this value out, and also populate the value later.
+    value: Option<ValueT>,
+}
+
+impl<ValueT> Item<ValueT> {
+    fn new(key: Key, value: Option<ValueT>) -> Self {
+        Self { key, value }
+    }
+}
+
+type ItemIndex = usize;
+
+#[derive(Debug)]
+pub struct AssociatedDataItemIndex {
+    index: ItemIndex,
+}
+impl AssociatedDataItemIndex {
+    fn new(index: ItemIndex) -> Self {
+        Self { index }
+    }
+}
+
+// A map from AST node to `T`, to assocaite extra data to AST node.
+#[derive(Debug)]
+pub struct AssociatedData<ValueT> {
+    next_index: ItemIndex,
+    items: Vec<Item<ValueT>>,
+    map: HashMap<Key, ItemIndex>,
+}
+
+impl<ValueT> AssociatedData<ValueT> {
+    pub fn new() -> Self {
+        Self {
+            next_index: 0,
+            items: Vec::new(),
+            map: HashMap::new(),
+        }
+    }
+
+    // Insert an item for `node`.
+    pub fn insert<NodeT>(&mut self, node: &NodeT, value: ValueT)
+    where
+        NodeT: SourceLocationAccessor + NodeTypeIdAccessor,
+    {
+        let index = self.items.len();
+        let key = Key::new(node);
+        self.items.push(Item::new(key.clone(), Some(value)));
+        self.map.insert(key, index);
+    }
+
+    // Allocate an item space for `node`, to populate it later by `populate`.
+    pub fn allocate<NodeT>(&mut self, node: &NodeT) -> AssociatedDataItemIndex
+    where
+        NodeT: SourceLocationAccessor + NodeTypeIdAccessor,
+    {
+        let index = self.items.len();
+        let key = Key::new(node);
+        self.items.push(Item::new(key.clone(), None));
+        self.map.insert(key, index);
+
+        AssociatedDataItemIndex::new(index)
+    }
+
+    // Populate the item allocated by `allocate`.
+    pub fn populate(&mut self, index: AssociatedDataItemIndex, value: ValueT) {
+        // NOTE: This can panic if `index` is created by another instance of
+        // `AssociatedData`.
+        self.items[index.index].value = Some(value);
+    }
+
+    // Get the item for `node`, removing it from this map.
+    // `None` if there's no item inserted for `node`, or it's already removed.
+    //
+    // This is O(1) if nodes are removed in the same order as `insert` or
+    // `allocate`.
+    pub fn remove<NodeT>(&mut self, node: &NodeT) -> Option<ValueT>
+    where
+        NodeT: SourceLocationAccessor + NodeTypeIdAccessor,
+    {
+        let key = Key::new(node);
+
+        // First, check the next item after the last call.
```

Comment:
Is this optimization necessary? I expect removes to be rare (and hash table lookups average out to constant-time anyway, though much slower).

It seems like you probably have a specific use case in mind, like removing a range of entries at once as part of transferring ownership of some parts from the AST to another data structure. It's fine to leave this in if you expect to use it.

Reply:
currently, `remove` is the only way to get data from this structure.
so, it's not for removing unnecessary data, but for consumer of this structure.

maybe it's better adding (or replacing with?) a method that returns possibly-mutable reference?

Reply:
looks like returning immutable reference works for my WIP patch for scope handling.
I'll replace `remove`  with `get`.